b2sdk/session.py

```diff
@@ -30,13 +30,31 @@ def __getattr__(self, name):
         @functools.wraps(f)
         def wrapper(*args, **kwargs):
             auth_failure_encountered = False
+            token_type = getattr(f, 'token_type', TokenType.API)
             # download_by_name uses different URLs
             url_factory = kwargs.pop('url_factory', self._api.account_info.get_api_url)
             while 1:
-                api_url = url_factory()
-                account_auth_token = self._api.account_info.get_account_auth_token()
                 try:
-                    return f(api_url, account_auth_token, *args, **kwargs)
+                    if token_type in [TokenType.UPLOAD_SMALL, TokenType.UPLOAD_PART]:
+                        bucket_id = args[0]
```

Comment:
what if someone passes the bucket_id via kwargs?
The `if` statement below kind of shows two different functions with a (very small) comman part that should be isolated to their own functions where we could use python's variable unpacking, so:
```python
if token_type == TokenType.API:
    account_auth_token = self._api.account_info.get_account_auth_token()
    return f(api_url, account_auth_token, *args, **kwargs)
elif token_type == TokenType.UPLOAD_SMALL:
    return self._upload_small(f, *args, **kwargs)
elif token_type == TokenType.UPLOAD_PART:
    return self._upload_part(f, *args, **kwargs)

# and then

def _upload_small(self, f, bucket_id, file_id, *args, **kwargs):
    upload_url, upload_auth_token = self._get_upload_data(bucket_id)
    response = f(upload_url, upload_auth_token, *args, **kwargs)
    self._api.account_info.put_bucket_upload_url(
        file_id, upload_url, upload_auth_token
    )
    return response
```